File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxlen=0
        hsh = {}
        left = 0
        n = len(s)

        for right in range(n):
            hsh[s[right]] = hsh.get(s[right],0)+1

            while hsh.get(s[right]) > 1:
                hsh[s[left]]-=1
                if hsh[s[left]] == 0:
                    del hsh[s[left]]
                left+=1
            
            maxlen = max(maxlen,right-left+1)

        return maxlen

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        Map<Character,Integer> hsh = new HashMap<>();
        int max_val = 0;

        int left = 0;

        for (int right = 0 ; right < s.length();right++){
            char ch = s.charAt(right);
            hsh.put(ch,hsh.getOrDefault(ch,0)+1);

            while (hsh.get(ch) > 1){
                char ch1 = s.charAt(left);
                hsh.put(ch1,hsh.get(ch1) - 1);
                
                if (hsh.get(ch1) == 0){
                    hsh.remove(ch1);
                }
                //hsh.put(s[left],get(l) - 1);
                left++;
            }

            max_val = Math.max(max_val,right-left+1);
        }

    return max_val;
        
    }
}